### PR TITLE
refactor(connector): [noon] update and add recommended fields 

### PR DESCRIPTION
**Pull request by SamraatBansal**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [X] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

- Add recommended fields by Noon like `BillingAddress`, `ipAddress`, `cancellationReason`

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

Screenshot 2023-09-27 at 5 32 15 PM

Screenshot 2023-09-27 at 5 33 04 PM

Screenshot 2023-09-27 at 5 33 18 PM

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [X] I reviewed the submitted code
- [X] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: config/config.example.toml [hunk 1]
```diff
@@ -187,6 +187,7 @@ multisafepay.base_url = "https://testapi.multisafepay.com/"
 nexinets.base_url = "https://apitest.payengine.de/v1"
 nmi.base_url = "https://secure.nmi.com/"
 noon.base_url = "https://api-test.noonpayments.com/"
+noon.key_mode = "Test"
```
Comment (ArjunKarthik) [inline config/config.example.toml:190]: You can consume test_mode in Router data for this use case

File: crates/router/src/configs/settings.rs [hunk 2]
```diff
@@ -552,6 +552,15 @@ pub struct ConnectorParams {
     pub secondary_base_url: Option<String>,
 }
 
+#[derive(Debug, Deserialize, Clone, Default, router_derive::ConfigValidate)]
+#[serde(default)]
+pub struct ConnectorParamsWithModeType {
```
Comment (ArjunKarthik) [inline crates/router/src/configs/settings.rs:557]: This struct is not required

File: crates/router/src/connector/noon/transformers.rs [hunk 3]
```diff
@@ -37,6 +39,23 @@ pub struct NoonSubscriptionData {
     name: String,
 }
 
+#[derive(Debug, Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct NoonBillingAddress {
+    street: Option<Secret<String>>,
+    street_2: Option<Secret<String>>,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/noon/transformers.rs:46]: street2

File: crates/router/src/connector/noon.rs [hunk 4]
```diff
@@ -58,22 +57,49 @@ impl
     // Not Implemented (R)
 }
 
+fn get_auth_header(
```
Comment (ArjunKarthik) [inline crates/router/src/connector/noon.rs:60]: Keep this inside ConnectorCommon itself, Since &settings::Connectors is no more required

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/noon.rs (hunk 4 above)
Comment (SamraatBansal) [reply crates/router/src/connector/noon.rs:60]: We would need to consume test_mode from router_data so we can't keep it inside ConnectorCommon

**Review by ArjunKarthik — APPROVED** [review]

**Review by Narayanbhat166 — APPROVED** [review]

**Pull request merged**